sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+inside explicit SIMD kernels and functions. Most of them will be eventually
+dropped. What's not supported today:
+- Mixing `EXPLICIT_SIMD` kernels with SYCL kernels in a single source
+- Local accessors. Local memory is allocated and accessed via explicit
+device-side API
+- 2D and 3D accessors
+- Constant accessors
+- `sycl::accessor::get_pointer()`. All memory accesses through an accessor are
+done via explicit APIs; e.g. `sycl::intel::gpu::block_store(acc, offset)`
+- Few others (to be documented)
+
+
+## Core Explicit SIMD programming APIs
+
+The DPC++ Explicit SIMD library defines the following classes to enhance the
+expressiveness for explicit SIMD data-parallel programming while enabling
+efficient mapping to SIMD vector operations on Intel GPU architectures.
+
+### SIMD vector class
+
+The ```sycl::intel::gpu::simd``` class is defined as a templated class for a vector
+of fundamental data elements. The fundamental data element type **__Ty** must be
+vectorizable, which can be any of the C++ character/integer/real floating-point types,
+or IEEE-754 half-precision floating-point type. The length of the vector
+is a constant expression, as determined by the template parameter **__N** in a
+given specialization.
+
+Each simd class object is mapped to a consecutive block of general register
+files (GRF).
+
+```cpp
+namespace cm {
+namespace gen {
+template <typename __Ty, int __N> class simd {
+public:
+  using value_type = simd<__Ty, __N, 0>;
+  using element_type = __Ty;
+  using vector_type = __vector_type_t<__Ty, __N>;
+  using type = simd<__Ty, __N>;
+  static constexpr int length = __N;
+
+  // Constructors.
+  constexpr simd();
+  constexpr simd(const simd &other);
+  constexpr simd(simd &&other);
+  constexpr simd(const vector_type &__Val);
+  constexpr simd(std::initializer_list<__Ty> __Ilist) noexcept;
+  constexpr simd(__Ty __Val, __Ty __Step = __Ty()) noexcept;
+
+  // Assignment operators.
+  constexpr type &operator=(const type &) &;
+  constexpr type &operator=(type &&) &;
+
+  // Subscript operators.
+  __Ty operator[](int __i);
+  __Ty operator[](int __i) const;
+
+  // Unary operators.
+  type &operator++();
+  type operator++(int);
+  type &operator--();
+  type operator--(int);
+
+  // Binary operators.
+  auto operator +(const type &__RHS) const;
+  auto operator -(const type &__RHS) const;
+  auto operator *(const type &__RHS) const;
+  auto operator /(const type &__RHS) const;
+  type operator &(const type &__RHS) const;
+  type operator |(const type &__RHS) const;
+  type operator ^(const type &__RHS) const;
+
+  // Compound assignment.
+  type &operator +=(const type &__RHS);
+  type &operator -=(const type &__RHS);
+  type &operator *=(const type &__RHS);
+  type &operator /=(const type &__RHS);
+  type &operator &=(const type &__RHS);
+  type &operator |=(const type &__RHS);
+  type &operator ^=(const type &__RHS);
+
+  // Compare operators.
+  simd<uint16_t, __N> operator >(const type &__RHS) const;
```

Comment:
This return type is very weird. Why is there no `simd_mask` type? Why is the storage format of the mask not implementation defined?

Reply:
In CM implementation simd mask type is mapped to vector of unsigned short.

Reply:
But why expose that to a user?

Reply:
added TODO to the headers